java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java

```diff
@@ -4569,6 +4614,57 @@ public int schedulePackageRemoveByNevra(User loggedInUser, final Integer sid, Li
                 ActionFactory.TYPE_PACKAGES_REMOVE, allowModules)[0].intValue();
     }
 
+    /**
+     * Schedule package lock for a system.
+     *
+     * @param loggedInUser The current user
+     * @param sid ID of the server
+     * @param pkgIdsToLock List of package IDs to lock (as Integers)
+     * @param pkgIdsToUnlock List of package IDs to lock (as Integers)
+     * @param earliestOccurrence Earliest occurrence of the package install
+     * @return package action id
+     *
+     * @xmlrpc.doc Schedule package installation for a system.
+     * @xmlrpc.param #param("string", "sessionKey")
+     * @xmlrpc.param #param("int", "serverId")
+     * @xmlrpc.param #array_single("int", "pkgIdsToLock")
+     * @xmlrpc.param #array_single("int", "pkgIdsToUnlock")
+     * @xmlrpc.param dateTime.iso8601 earliestOccurrence
+     * @xmlrpc.returntype #param_desc("return_int_success")
+     */
+    public Long schedulePackageLockChange(User loggedInUser, Long sid,
+                                          List<Long> pkgIdsToLock,
+                                          List<Long> pkgIdsToUnlock, Date earliestOccurrence) {
+
+        Server server = SystemManager.lookupByIdAndUser(sid.longValue(), loggedInUser);
+        DataResult<PackageListItem> lockedPackagesResult =
+                PackageManager.systemLockedPackages(server.getId(), null);
+
+        Set<Package> pkgsAlreadyLocked = lockedPackagesResult.stream().map(
+                pkg -> PackageManager.lookupByIdAndUser(pkg.getPackageId(), loggedInUser)).collect(Collectors.toSet());
+        Set<Package> pkgsToLock = pkgIdsToLock.stream().map(
+                pkg -> PackageManager.lookupByIdAndUser(pkg.longValue(), loggedInUser)).
+                filter(Objects::isNull).collect(Collectors.toSet());
```

Comment:
Not sure to follow the logic here, maybe I'm missing something. Why are you collecting the ```null``` elements? Shouldn't it be ```Objects::nonNull```? Anyway, I think in this case you can avoid traversing the stream twice by using ```peek```:

```
pkgIdsToLock.stream().map( ... ).filter( ... ).peek(pkg -> pkg.setLockPending(Boolean.TRUE)).collect(Collectors.toSet())
```